Change homepage and middleclick function on links

Dimi20cen opened the pull request:
Regarding the homepage's search box, I saw the discussion that keeping the search-box round is important. However index.html is not very readable, at least for me ( i dont really do html/css/js ) so I can't find the reason that makes the search-box so "stubborn". If you think it's a good idea, I'm open to rewrite the code for the index file from the ground up, that might be more fun.

quozl commented:
What are _link1_ to _link12_ actually for?  If no reason, why are you maintaining them rather than removing them?

What does the variable _maintain_view_ mean?

quozl commented:
Thanks.  So by `maintain_view` you meant "do not change what is currently displayed, but rather maintain the view as is".  Using this boolean gives confusing code.  Made worse by the `next_to_current` changes made previously.  Please refactor, for example;
* make an `open_new_tab_background` method, and a `new-tab-background` signal,
* move the `focus-url-entry` emit from `add_tab` into the callers.

Dimi20cen commented:
So you suggest removing the maintain_view entirely?

Dimi20cen commented:
I made the suggested changes(removed main_view entirely and moved the focus-url-entry emit to the "__new_tab_cb"), they seem to work. Should I edit the previously commited files inside this pull request?

quozl commented on browser.py:
```diff
@@ -243,9 +243,23 @@ def __page_removed_cb(self, notebook, child, pagenum):
         new_browser.grab_focus()
 
+    def __new_tab_background_cb(self, browser, url):
+        browser = Browser(self._activity)
```
To be consistent with `__new_tab_cb` and not confuse with method parameter of the same name, local variable `browser` could be `new_browser`.

quozl commented on browser.py:
```diff
@@ -243,9 +243,23 @@ def __page_removed_cb(self, notebook, child, pagenum):
+        next_index = self.get_current_page() + 1
+        self.insert_page(tab_page, label, next_index)
+        tab_page.show()
+        self.set_current_page(next_index - 1)
```
Some of this code is similar to `add_tab` and what it calls.  The common code could be in a new method called by `add_tab` and `__new_tab_background_cb`.

Also, some of the code in `add_tab` is not included, such as connecting the signals.

quozl commented on data/index.html:
```diff
@@ -15,35 +15,35 @@
             .center {text-align: center; position: static; border: 0}
-            .search{background: #FFFFFF; border-radius: 25px; border-width: 1px; border-color: #686868; height: 32px; width: 45%; outline: 0; padding: 0; padding-left: 5px; padding-right: 5px; margin: 0}
-            .btn{background: #5A89EB; border-radius: 25px; width: 15%; height: 32px; color:#FFFFFF; font-style: normal; border-color: #FFFFFF; padding:0; margin: 0; margin-left: 5px; outline: 0; cursor: pointer}
+            .search{background: #0000; border-radius: 25px; border-width: 1px; border-color: #686868; height: 32px; width: 45%; outline: 0; padding: 0; padding-left: 5px; padding-right: 5px; margin: 0;}
```
Style is to use six digit hexadecimal colours or names.  You introduce a four digit colour.  Should you use `black`?

Dimi20cen replied:
To be honest, I thought that using a four-digit code for black wouldn't be an issue. I chose black because even when I tried other colors like white or blue (e.g., 4285F4), the issue persisted. As I mentioned in my pr message, this "solution" I provided for the search-box is more of a temporary workaround, rather than a perfect fix, since I don't understand the reason behind the missing corners.

quozl commented on data/index.html:
```diff
@@ -15,35 +15,35 @@
-            .link1 { padding-bottom: 4px; padding-top: 4px; text-decoration: none; color: #006e00;}
+            html { margin-right: 8px; margin-left: 8px; font-family: Helvetica, Arial, sans-serif;}
+            .links-container { padding-bottom: 4px; padding-top: 4px; display: block; text-align: right;}
+            .link1 { padding:8px; text-decoration: none; color: #006e00; background-color: rgba(211, 211, 211, 0.2); border-radius: 10px; transition: background-color 0.5s;}
```
There is a lot of redundant coding here; can the common attributes be moved to a style that is inherited, so that `.link1` and `.link2` differ only in `color`?

quozl reviewed (changes requested):
Thanks.  Looking better.

quozl commented:
Thanks.  Some functional tests were done.

* [ ] the search entry box has corners that are not rendered clearly; there are rectilinear chunks missing,
* [ ] the new tab is always created immediately to the right of the current tab, which means that multiple actions are in reverse order in the tabs,
* [ ] right-click on a PDF link then _Follow link in new tab_ results in an _Untitled_ new tab, **and** another tab with the PDF rendered, (web page used for test was https://wiki.sugarlabs.org/go/Marketing_Team/Screen_Shots)
* [ ] using the new tab button on the tab bar, or the Control/T shortcut, did not place focus on the URL entry box,
* [ ] middle mouse button on a tab title did not close the tab,
* [ ] there are no shortcuts for switching tabs,
* [ ] there are no drag methods for reordering tabs,
* [ ] during testing some strange messages were printed on stderr,

```log
(process:1252): GLib-CRITICAL **: 17:21:21.803: Failed to set scheduler settings: Operation not permitted
Traceback (most recent call last):
  File "/usr/share/sugar/activities/Browse.activity/pdfviewer.py", line 567, in __download_finished_cb
    self._show_pdf()
  File "/usr/share/sugar/activities/Browse.activity/pdfviewer.py", line 494, in _show_pdf
    self._evince_viewer = EvinceViewer(self._pdf_uri)
  File "/usr/share/sugar/activities/Browse.activity/pdfviewer.py", line 66, in __init__
    self._doc = EvinceDocument.Document.factory_get_document(uri)
gi.repository.GLib.Error: g-io-error-quark: Error opening file /home/guest/Downloads/Screenshot_of_0.84_home.pdf: No such file or directory (1)
/usr/lib/python3/dist-packages/gi/overrides/GObject.py:502: Warning: ../../../gobject/gsignal.c:2736: instance '0x24d3e40' has no handler with id '8493'
  return func(*args, **kwargs)
Traceback (most recent call last):
  File "/usr/share/sugar/activities/Browse.activity/edittoolbar.py", line 130, in __can_execute_editing_command_cb
    can = self._browser.can_execute_editing_command_finish(result)
TypeError: can_execute_editing_command_finish() takes 1 positional argument but 2 were given
```

Dimi20cen commented:
How should I proceed with regards to making new commits? Should I keep making new commits to solve further issues?

quozl commented:
That's up to you.  You could either add new commits referencing the review and the commits that relate to it, or rebase and rewrite commits.  The possibilities are somewhat extensive.  If it were me, I'd reorder the commits so that the refactoring happens first, because that refactoring is unlikely to be contentious and so could be merged early, and _also_ leave the resulting git history in a useful state for `git bisect`.  It is quite annoying to be bisecting a problem and having history at a point in time not executable or quite different.  Your commit messages are quite good.  Put yourself in the minds of your reviewers.  Read our checklist for reviewers.  Try the review process yourself.

Other options are to convert the pull request to draft so as to reduce our engagement and give yourself time to go into more detail, or call other developers to review and comment, such as @chimosky.

Dimi20cen commented:
Thank you for the feedbacks and for taking the time.
I have tried to clean up the commit history to make it more accessible and easier to work with. I dropped the 5 previous commits and created 4 new, that are better ordered. The initial commit refactors the ```index.html``` and the subsequent commits are structured in stages to help the review process.
Hopefully those changes, increase convenience for reviewing, if you have additional feeback please let me know.

Also regarding the middle-clicking
- I corrected the tab placement, so now the new tab is added at the last index. And using the new tab button on the tab bar, or the CTRL+T shortcut now places focus on the URL entry box.
- However, the issue with the pdf links persists. The cause for the issue is not the change of function of the middle-click, since the same issue is present in the current version of the browser as well and is present with whichever way the user tries to open a pdf url.

quozl commented:
a8e289126bda91d2be6282521522685eebfad2ff was merged with commit message changes.

quozl commented:
e8528a5f83 was merged with commit message changes.

quozl commented:
I disagree with the changes in [a877389](https://github.com/sugarlabs/browse-activity/pull/124/commits/a87738931160e8f75aa5ec6230b81adf736ce4ab) because it is important to ensure children learn that coloured text represents a link to another document.  It doesn't matter that it looks outdated, we're not selling this as a service.

quozl commented:
efe02bd has broken middle-click to open links in background tab.

chimosky commented on data/index.html:
```diff
@@ -111,16 +112,13 @@
-                { href: 'https://wiki.sugarlabs.org/', text: 'wiki' },
-                { href: 'http://schoolserver/', text: 'school' },
-                { href: 'https://v4.activities.sugarlabs.org/', text: 'activities' }
+                { href: 'https://www.sugarlabs.org/', text: 'Sugar labs' },
```
It's Sugar Labs.

Dimi20cen replied:
Yep Sugar Labs it is

chimosky commented:
Reviewed efe02bd, not tested.

Dimi20cen commented:
> I disagree with the changes in [a877389](https://github.com/sugarlabs/browse-activity/pull/124/commits/a87738931160e8f75aa5ec6230b81adf736ce4ab) because it is important to ensure children learn that coloured text represents a link to another document. It doesn't matter that it looks outdated, we're not selling this as a service.

Yep I agree, that's why I didnt change the link color for links, the change was regarding the background of the links when hovered and the removal of the "/".

Dimi20cen commented:
@chimosky, in the Matrix channel, I was asking about the 2cars-activity.

As for this PR, I believe it's intended goal has been accomplished with the latest commits, provided I haven't missed something. Also in the case of closing empty tabs and in some other cases, which I can't recall ( I have to get better at keeping notes ), the warnings can also be replicated with the existing browse-activity.

chimosky commented:
Merged by hand, thank you!
